crates/noirc_frontend/src/hir/type_check/errors.rs

```diff
@@ -117,48 +97,32 @@ impl TypeCheckError {
                     first_span,
                 );
                 diag.add_secondary(format!("but then found type {}", second_type), second_span);
-                vec![diag]
+                diag
             }
             TypeCheckError::ArityMisMatch {
                 expected,
                 found,
                 span,
-            } => {
-                vec![Diagnostic::simple_error(
-                    format!("expected {} number of arguments, found {}", expected, found),
-                    String::new(),
-                    span,
-                )]
-            }
+            } => Diagnostic::simple_error(
+                format!("expected {} number of arguments, found {}", expected, found),
+                String::new(),
+                span,
+            ),
             TypeCheckError::Unstructured { msg, span } => {
-                vec![Diagnostic::simple_error(msg, String::new(), span)]
-            }
-            TypeCheckError::PublicReturnType { typ, span } => {
-                vec![Diagnostic::simple_error(
-                    "functions cannot declare a public return type".to_string(),
-                    format!("return type is {}", typ),
-                    span,
-                )]
+                Diagnostic::simple_error(msg, String::new(), span)
             }
+            TypeCheckError::PublicReturnType { typ, span } => Diagnostic::simple_error(
+                "functions cannot declare a public return type".to_string(),
+                format!("return type is {}", typ),
+                span,
+            ),
         }
     }
 
-    pub fn add_context(self, ctx: &'static str) -> Option<Self> {
-        match &self {
-            TypeCheckError::OpCannotBeUsed { .. }
-            | TypeCheckError::Unstructured { .. }
-            | TypeCheckError::TypeMismatch { .. }
-            | TypeCheckError::NonHomogeneousArray { .. }
-            | TypeCheckError::PublicReturnType { .. }
-            | TypeCheckError::ArityMisMatch { .. }
-            | TypeCheckError::TypeCannotBeUsed { .. } => Some(TypeCheckError::Context {
-                err: Box::new(self),
-                ctx,
-            }),
-            // Cannot apply a context to multiple diagnostics
-            TypeCheckError::MultipleErrors(_) => None,
-            // Cannot append or overwrite a context
-            TypeCheckError::Context { .. } => None,
+    pub fn add_context(self, ctx: &'static str) -> Self {
+        TypeCheckError::Context {
+            err: Box::new(self),
+            ctx,
```

Comment:
This is a functional change actually, it lets us add multiple contexts if necessary. Was there a reason this was disallowed to begin with? It seems useful to add multiple contexts as long as we can avoid confusing the user.

Reply:
Yeah it was mainly to not confuse the reader and I did not initially see the reason to do it. I don't think there is anything wrong with adding it though